src/Compilers/Core/CommandLine/BuildClient.cs

```diff
@@ -149,36 +136,65 @@ public Task<RunCompilationResult> RunCompilationAsync(IEnumerable<string> origin
 
         protected abstract int RunLocalCompilation(string[] arguments, BuildPaths buildPaths, TextWriter textWriter);
 
-        protected abstract Task<BuildResponse> RunServerCompilation(List<string> arguments, BuildPaths buildPaths, string sessionName, string keepAlive, string libDirectory, CancellationToken cancellationToken);
-
-        protected abstract string GetSessionKey(BuildPaths buildPaths);
-
-        protected virtual RunCompilationResult HandleResponse(BuildResponse response, string[] arguments, BuildPaths buildPaths, TextWriter textWriter)
+        /// <summary>
+        /// Runs the provided compilation on the server.  If the compilation cannot be completed on the server then null
+        /// will be returned.
+        /// </summary>
+        internal RunCompilationResult? RunServerCompilation(TextWriter textWriter, List<string> arguments, BuildPaths buildPaths, string libDirectory, string sessionName, string keepAlive)
         {
-            switch (response.Type)
+            BuildResponse buildResponse;
+
+            try
             {
-                case BuildResponse.ResponseType.MismatchedVersion:
-                    Console.Error.WriteLine(CommandLineParser.MismatchedVersionErrorText);
-                    return RunCompilationResult.Failed;
+                var buildResponseTask = RunServerCompilation(
+                    arguments,
+                    buildPaths,
+                    sessionName,
+                    keepAlive,
+                    libDirectory,
+                    CancellationToken.None);
+                buildResponse = buildResponseTask.Result;
+
+                Debug.Assert(buildResponse != null);
+                if (buildResponse == null)
+                {
+                    return null;
+                }
+            }
+            catch (Exception)
```

Comment:
I did give this some consideration and looking through Roslyn it seems we are a bit inconsistent with our `catch` usage in this area.  In general it shouldn't be a problem since the runtime defaults to converting objects to an exception type.  In theory though we could be run in an environment which didn't wrap object in exception and actually threw one via an extensibility point. Eventually I decided that was a bit extreme.  Willing to be convinced otherwise.